### ISSUE TITLE
AI-Powered Log Analysis and Automated Troubleshooting

### ISSUE THREAD

**Issue by makigjuro** [signalbeam-io/signalbeam-edge#220]
# AI-Powered Log Analysis and Automated Troubleshooting

## Overview
Use AI to automatically analyze device logs, detect patterns, identify root causes, and suggest/execute remediation actions, reducing MTTR (Mean Time To Resolution).

## Business Value
- **Reduce MTTR:** From hours to minutes
- **Scale operations:** One operator can manage 1000+ devices
- **Reduce costs:** 80% fewer escalations to engineering
- **Improve reliability:** Auto-fix common issues

### 1. Automatic Root Cause Analysis
**Problem:** Device offline, but why?

**AI Solution:**
- Aggregate logs from device, network, containers
- Apply pattern matching + LLM analysis
- Identify: "Network timeout due to router firmware bug"
- Suggest: "Reboot router or update firmware"

### 2. Log Pattern Recognition
**Problem:** Thousands of log lines, hard to find issues

**AI Solution:**
- Extract log patterns using clustering
- Identify anomalous patterns
- Surface: "New error pattern appeared 30 min ago: 'Database connection timeout'"
- Correlate with deployments/changes

### 3. Intelligent Error Correlation
**Problem:** Multiple devices showing same error

**AI Solution:**
- Detect similar error patterns across fleet
- Group by root cause
- Alert once: "15 devices failing with DNS resolution error - likely ISP issue"

### 4. Auto-Remediation
**Problem:** Common issues require manual intervention

**AI Solution:**
- Recognize known error patterns
- Execute remediation playbook
- Example: "Disk full → Clean up old logs → Verify → Alert resolved"

### Components

**1. Log Ingestion Pipeline**
```
Edge Agent → NATS → Log Aggregator → TimescaleDB → AI Analysis Engine
```

**2. AI Analysis Engine**
```
Logs → Preprocessing → Pattern Extraction → LLM Analysis → Actions
                                                ↓
                                         Knowledge Base
```

### Pattern Extraction

```python
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.cluster import DBSCAN

class LogPatternExtractor:
    def extract_patterns(self, logs: List[str]) -> List[Pattern]:
        # Remove timestamps, IDs, etc.
        normalized = [self.normalize_log(log) for log in logs]
        
        # Vectorize logs
        vectorizer = TfidfVectorizer(max_features=1000)
        vectors = vectorizer.fit_transform(normalized)
        
        # Cluster similar logs
        clustering = DBSCAN(eps=0.3, min_samples=5)
        labels = clustering.fit_predict(vectors)
        
        # Extract representative pattern for each cluster
        patterns = []
        for cluster_id in set(labels):
            if cluster_id == -1:  # Noise
                continue
            
            cluster_logs = [logs[i] for i, l in enumerate(labels) if l == cluster_id]
            pattern = self.extract_template(cluster_logs)
            patterns.append(Pattern(
                template=pattern,
                frequency=len(cluster_logs),
                severity=self.classify_severity(pattern)
            ))
        
        return patterns
```

### LLM Integration for Root Cause Analysis

```python
import anthropic

class AITroubleshooter:
    def analyze_issue(self, device_id: str, symptoms: str) -> TroubleshootingResult:
        # Gather context
        logs = fetch_device_logs(device_id, hours=1)
        metrics = fetch_device_metrics(device_id, hours=1)
        recent_changes = fetch_recent_changes(device_id)
        
        # Construct prompt
        prompt = f"""
        Analyze this device issue and provide troubleshooting steps.
        
        Device: {device_id}
        Symptoms: {symptoms}
        
        Recent Logs (last 100 lines):
        {logs}
        
        Recent Metrics:
        CPU: {metrics.cpu}%
        Memory: {metrics.memory}%
        Disk: {metrics.disk}%
        
        Recent Changes:
        {recent_changes}
        
        Provide:
        1. Root cause analysis
        2. Immediate remediation steps
        3. Long-term prevention
        4. Confidence level (0-100)
        """
        
        response = self.claude.messages.create(
            model="claude-sonnet-4-5",
            max_tokens=2000,
            messages=[{"role": "user", "content": prompt}]
        )
        
        return self.parse_response(response.content[0].text)
```

### Auto-Remediation Engine

```python
class RemediationEngine:
    """Execute automated fixes for known issues"""
    
    playbooks = {
        "disk_full": [
            {"action": "cleanup_old_logs", "params": {"days": 7}},
            {"action": "remove_unused_images", "params": {}},
            {"action": "verify_disk_space", "threshold": 80}
        ],
        "memory_leak": [
            {"action": "restart_container", "params": {"graceful": True}},
            {"action": "verify_container_health", "timeout": 60},
            {"action": "monitor_memory", "duration": 300}
        ],
        "network_timeout": [
            {"action": "check_connectivity", "target": "8.8.8.8"},
            {"action": "restart_network_interface", "if_failed": True},
            {"action": "verify_connectivity", "timeout": 30}
        ]
    }
    
    async def execute_playbook(self, playbook_name: str, device_id: str):
        steps = self.playbooks[playbook_name]
        results = []
        
        for step in steps:
            action = step["action"]
            params = step.get("params", {})
            
            result = await self.execute_action(device_id, action, params)
            results.append(result)
            
            if not result.success and step.get("if_failed"):
                break
        
        return RemediationResult(
            playbook=playbook_name,
            steps=results,
            overall_success=all(r.success for r in results)
        )
```

### Log Streaming Architecture

```typescript
// Real-time log streaming from agents
class LogStreamer {
  async streamLogs(deviceId: string): AsyncIterator<LogEntry> {
    const subject = `signalbeam.devices.${deviceId}.logs`;
    const subscription = await this.nats.subscribe(subject);
    
    for await (const msg of subscription) {
      const logEntry = JSON.parse(msg.data);
      
      // Parse and enrich
      const parsed = this.parseLog(logEntry);
      const enriched = await this.enrichLog(parsed);
      
      // Store in TimescaleDB
      await this.storeLog(enriched);
      
      // Check for anomalies
      if (await this.isAnomalous(enriched)) {
        await this.triggerAnalysis(deviceId, enriched);
      }
      
      yield enriched;
    }
  }
}
```

### Knowledge Base

```sql
CREATE TABLE LogPatterns (
    Id UUID PRIMARY KEY,
    Pattern TEXT NOT NULL,
    PatternType VARCHAR(50) NOT NULL,  -- error, warning, info
    Severity VARCHAR(20) NOT NULL,
    RootCause TEXT,
    RemediationPlaybook VARCHAR(100),
    Frequency INT DEFAULT 0,
    LastSeen TIMESTAMP,
    CreatedAt TIMESTAMP NOT NULL,
    INDEX idx_log_patterns_type (PatternType)
);

CREATE TABLE RemediationHistory (
    Id UUID PRIMARY KEY,
    DeviceId UUID NOT NULL REFERENCES Devices(Id),
    Issue VARCHAR(100) NOT NULL,
    Playbook VARCHAR(100) NOT NULL,
    ExecutedAt TIMESTAMP NOT NULL,
    Success BOOLEAN NOT NULL,
    DurationSeconds INT,
    Actions JSONB,
    INDEX idx_remediation_history_device (DeviceId)
);
```

## API Endpoints

```python
# AI Analysis Service

@app.post("/api/ai/analyze-logs")
async def analyze_logs(device_id: str, hours: int = 1):
    logs = fetch_logs(device_id, hours=hours)
    patterns = pattern_extractor.extract_patterns(logs)
    analysis = ai_troubleshooter.analyze_patterns(patterns)
    
    return {
        "device_id": device_id,
        "patterns_found": len(patterns),
        "anomalies": analysis.anomalies,
        "recommendations": analysis.recommendations
    }

@app.post("/api/ai/troubleshoot")
async def troubleshoot(device_id: str, symptoms: str):
    result = ai_troubleshooter.analyze_issue(device_id, symptoms)
    
    return {
        "root_cause": result.root_cause,
        "confidence": result.confidence,
        "remediation_steps": result.steps,
        "auto_fix_available": result.playbook is not None
    }

@app.post("/api/ai/auto-fix")
async def auto_fix(device_id: str, issue_type: str, dry_run: bool = True):
    if dry_run:
        # Preview what would be done
        playbook = remediation_engine.get_playbook(issue_type)
        return {"preview": playbook, "dry_run": True}
    
    # Execute remediation
    result = await remediation_engine.execute_playbook(issue_type, device_id)
    
    return {
        "success": result.overall_success,
        "steps_executed": len(result.steps),
        "duration_seconds": result.duration
    }
```

### Log Viewer with AI Insights
```typescript
function LogViewer({ deviceId }: { deviceId: string }) {
  const { data: analysis } = useQuery({
    queryKey: ['log-analysis', deviceId],
    queryFn: () => api.ai.analyzeLogs(deviceId)
  });
  
  return (
    <div>
      {analysis?.anomalies.length > 0 && (
        <Alert severity="warning">
          <AlertTitle>AI Detected Anomalies</AlertTitle>
          {analysis.anomalies.map(a => (
            <div key={a.id}>
              <strong>{a.pattern}</strong>: {a.description}
              <Button onClick={() => triggerAutoFix(a)}>
                Auto-Fix
              </Button>
            </div>
          ))}
        </Alert>
      )}
      
      <LogStream deviceId={deviceId} />
    </div>
  );
}
```

## Configuration

```json
{
  "AI": {
    "LogAnalysis": {
      "Enabled": true,
      "AnalysisIntervalMinutes": 15,
      "AnomalyThreshold": 0.8,
      "PatternMinOccurrences": 5
    },
    "AutoRemediation": {
      "Enabled": false,  // Require explicit enable
      "RequireApproval": true,
      "MaxConcurrentFixes": 10,
      "AllowedPlaybooks": ["disk_full", "container_restart"]
    },
    "Claude": {
      "ApiKey": "sk-...",
      "Model": "claude-sonnet-4-5",
      "MaxTokens": 2000
    }
  }
}
```

## Safety & Governance

- **Approval workflow:** Auto-fixes require approval for production
- **Dry-run mode:** Preview all actions before execution
- **Rollback capability:** Undo auto-fixes if they fail
- **Audit log:** Every AI action logged
- **Rate limiting:** Max 10 auto-fixes per device per day
- **Human oversight:** Alert on high-risk actions

## Success Metrics
- 80% of issues auto-diagnosed within 5 minutes
- 50% of common issues auto-fixed
- 70% reduction in MTTR
- 90% root cause accuracy
- <5% false positive auto-fixes

## Dependencies
- Claude API access
- NATS log streaming (#215)
- TimescaleDB for logs
- Python 3.11+ with sklearn, anthropic SDK

## Future Enhancements
- Multi-device correlation analysis
- Predictive issue detection
- Custom playbook builder UI
- Natural language playbook creation
- Federated learning from fleet
